File: Blind150/2.Two Pointer/Python/three_sum.py
# https://leetcode.com/problems/3sum/description/
#
class Solution:

    def twoSum(self, l,r, nums , target):

        ans = []

        while l < r:

            cur_sum = nums[l] + nums[r]
            if  cur_sum > target:

                r -= 1

            elif cur_sum < target:

                l += 1

            else:
                ans.append([nums[l],nums[r]])

                l += 1

                # dry run this [-2,0,0,2,2]
                while nums[l] == nums[l-1] and l < r:
                    l += 1


        return ans


    def threeSum(self, nums: List[int]) -> List[List[int]]:

        res = []
        nums.sort()


        for pos, val in enumerate(nums):

            if pos > 0 and nums[pos] == nums[pos-1]:
                continue

            l,r = pos+1, len(nums)-1

            found = self.twoSum(l,r,nums,0-val)

            for r in found:
                res.append([val]+ r)

        return res

File: Blind150/2.Two Pointer/GO/three_sum.go
// https://leetcode.com/problems/3sum/description/


import "sort"

func twoSum(l int,r int, nums []int, target int) [][]int {

    ans := [][]int{}
    for l < r {

        cur_sum := nums[l] + nums[r]

        if  cur_sum > target{

            r = r - 1
        }else if cur_sum < target{

            l = l + 1
        }else{

            ans  = append(ans, []int{nums[l],nums[r]})

            l = l + 1

            for nums[l] == nums[l-1] && l < r {

                l = l + 1
            }
        }
    }
    return ans
}

func threeSum(nums []int) [][]int {


        res := [][]int{}
        // To sort an integer slice:
        sort.Ints(nums)

        for pos, val := range nums{

            if pos > 0 && nums[pos] == nums[pos-1]{
                continue
            }
            l,r := pos+1, len(nums)-1

            found := twoSum(l,r,nums,0-val)

            for _, r := range found{

                // Create a new slice with val and append r
                res = append(res, append([]int{val}, r...))

            }
        }
        return res
}
